### PR TITLE
FABN-1534 Set couchdb security

### PR DESCRIPTION
**Pull request by xuxucode**
Should set security for couchdb wallet store when couchdb authentication is enabled, or else the documents are public.

For example, `curl http://localhost:5984/wallet/_all_docs`,
### Before setting security
```
{"total_rows":3,"offset":0,"rows":[
  {"id":"admin","key":"admin","value":{"rev":"2-67dcf1c6561c1d33c1705f8a0519a9d7"}},
  {"id":"user1","key":"company2","value":{"rev":"1-a773a678afbd685f13a7a3cd4442455e"}},
  {"id":"user2","key":"company3","value":{"rev":"1-d8b41146b973dec3c8acef4bfaa681f4"}}
]}
```

### After setting security
```
{"error":"unauthorized","reason":"You are not authorized to access this db."}
```

### CODE REVIEW

**Review by bestbeforetoday — COMMENTED** [review]
This really should have accompanying tests. There used to be a (tape) integration test for the CouchDB wallet that drove it against a real CouchDB. That seems to have been lost so I'll get an issue raised to restore that. I'm not sure how easy or practical it will be to test this security aspect though, which raises the question...

Should we be trying to create a CouchDB database _and_ configuring security for it on behalf of the user? Would we be better leaving security as something to configure themselves?

File: fabric-network/src/impl/wallet/couchdbwalletstore.ts [hunk 1]
```diff
@@ -21,6 +22,19 @@ export class CouchDBWalletStore implements WalletStore {
 			await client.db.get(dbName); // Throws if database does not exist
 		} catch (error) {
 			await client.db.create(dbName);
+			// Set security
```
[the remark this answers is not in the document]
Comment (xuxucode) [reply fabric-network/src/impl/wallet/couchdbwalletstore.ts]: I agree

File: fabric-network/src/impl/wallet/couchdbwalletstore.ts [hunk 2]
```diff
@@ -21,6 +22,19 @@ export class CouchDBWalletStore implements WalletStore {
...
+			const url = new URL(client.config.url);
+			if (url.username) {
+				await client.request({
```
[the remark this answers is not in the document]
Comment (xuxucode) [reply fabric-network/src/impl/wallet/couchdbwalletstore.ts]: Thanks, I've tried to solve it, but I am not familiar with TypeScript, could you help solve it?

**Comment by bestbeforetoday** [comment]
I think the problem is that the _nano_ TypeScript definition is wrong:
- https://github.com/apache/couchdb-nano/blob/f04b24fd1a874c629af6d0b27fce1e6978613cc1/lib/nano.d.ts#L36

This should be a function, not a property. Either the _nano_ TypeScript definitions needs to be corrected, or this code would need to work around the error by using a type assertion to persuade the compiler that you are doing the right thing despite the bug in the _nano_ TypeScript definitions:
- https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

I am not feeling comfortable merging this without some kind of test coverage to ensure it actually works.

**Comment by xuxucode** [comment]
I only tested it manually in my project, we need some tests

**Comment by davidkhala** [comment]
Would you mind creating an issue on JIRA alongside to track this?

**Comment by xuxucode** [comment]
> Would you mind creating an issue on JIRA alongside to track this?

FABN-1534 created

**Comment by xuxucode** [comment]
Maybe this issue can be closed, because in CouchDB 3.0.0 all databases are created by default as admin-only.

https://docs.couchdb.org/en/latest/whatsnew/3.0.html:
> All databases are now created by default as admin-only. That is, the default new database _security object is now:
{
  "members" : { "roles" : [ "_admin" ] },
   "admins" : { "roles" : [ "_admin" ] }
}

**Pull request closed without merging**